Integrate DEAL reconstruction model into DeepInverse

Holimmo7 opened the pull request:
This PR adds the full DEAL reconstruction model **inside DeepInverse**, so users no longer need to install the external `deal` package. The model is now fully integrated and follows the same structure as all other models in `deepinv.models`.

The goal is to make DEAL easy to use inside DeepInverse while keeping the original logic unchanged.

---

## Overview

The official DEAL implementation  
(*Mehrsa Pouya & Erich Kobler, “DEALing with Image Reconstruction”, 2025*)  
has been imported into the DeepInverse repository as an internal module (`deal_impl.py` and related files).

A wrapper class (`deepinv.models.DEAL`) provides the standard DeepInverse interface:

- behaves like built-in models (`forward(y, physics)` → reconstructed image)
- subclasses `Reconstructor`
- works with any physics operator implementing `__call__` and `A_adjoint`
- supports grayscale and color modes

---

## What Is Included

###  Internal model implementation  
Imported from the official DEAL source:

- `deepinv/models/deal_impl.py`  
- `deepinv/models/deal_linearspline.py`  
- `deepinv/models/deal_linearspline_utils.py`  
- `deepinv/models/deal_multi_conv.py`  
- `deepinv/models/deal_spline_autograd_func.py`  

###  Integration into DeepInverse API

- New wrapper: `deepinv/models/deal.py`  
- Added to public API via `deepinv/models/__init__.py`

###  Unit test

`deepinv/tests/test_deal_reconstructor.py`:

- injects a dummy DEAL model  
- uses a fake checkpoint  
- ensures the wrapper:
  - loads weights correctly  
  - runs `forward(y, physics)`  
  - returns a tensor of correct shape  

### Code style

All new files formatted using **black**.

---

## Important: Downloading Pretrained Checkpoints

DEAL's architecture is fully included, but **the pretrained weights are not stored inside this repository** to keep DeepInverse lightweight.

Users must manually download pretrained checkpoints from the official DEAL GitHub:

https://github.com/mehrsapo/DEAL/tree/main/trained_models

- `deal_gray.pth` (grayscale)
- `deal_color.pth` (RGB version)
---
## Example of usage

from deepinv.models import DEAL

model = DEAL(
    checkpoint_path="deal_gray.pth",
    sigma=25.0,
    lam=10.0,
    max_iter=50,
)

given physics operator H
x_hat = model(y, H)

---

Citation:

M. Pouya & E. Kobler
“DEALing with Image Reconstruction: Deep Attentive Least Squares”
arXiv:2502.04079, 2025
https://arxiv.org/abs/2502.04079

Official code:
https://github.com/mehrsapo/DEAL

jscanvic commented:
Thank you for making this PR!

Please go ahead, a short unit test will indeed suffice here

Holimmo7 commented:
Hi,

I have now refactored the implementation :

Integrated DEAL directly as a DeepInverse **model** under `deepinv/models/deal.py`
Removed the old `reconstructors/deal.py`
Updated `deepinv/models/__init__.py` and `deepinv/__init__.py`
Updated and simplified the unit test (`test_deal_reconstructor.py`) so it matches the new model structure
Verified that everything works by running `pytest` locally (the test passes)

Please let me know if you would like any further changes or improvements.

Thank you!
 Hossein

jscanvic commented:
Thanks for dealing with the CI failures! You might want to have a look at our contributing guide to deal with those linter failures

https://deepinv.github.io/deepinv/contributing.html

We're using two distinct tools `black` and `ruff` to check the code, you can run them locally to identify the failures:

- `black --check .`
- `ruff check`

You can also automatically fix most of the issues using

- `black .`
- `ruff check --fix` or `ruff check --fix --unsafe`

Hope that helps!

Tmodrzyk commented:
Hi @holimmo thank you for implementing the method into deepinverse ! I'll review the PR shortly :smile:

Tmodrzyk commented on deepinv/tests/test_deal_reconstructor.py:
Thank you very much for this thorough testing. 
This whole test should be included in ``deepinv/tests/test_models.py`` to make it more consistent with the other models. 
For the tests we usually use ``pytest``. Using the pytest fixtures in ``deepinv/tests/test_models.py`` will provide you with dummy measurements, dummy physics etc.
I think the best way to proceed would be to add an entry to the list passed to the ``test_restoration_models`` test in the ``test_models.py`` file.

Holimmo7 commented:
Dear @Tmodrzyk,
Thank you very much for the updates and improvements. 
Please let me know if anything else is needed from my side.

Tmodrzyk commented on examples/models/demo_deal.py:
```diff
@@ -0,0 +1,52 @@
+for a simple deblurring problem.
+"""
+
+import torch
```
We should start a new cell here with ``#%%``, with a title indicating what this step does (here, initialization and loading the butterfly image from Set3C)

matthieutrs commented:
Hi @Holimmo7, sorry about this wild revert!

I think we have some ways to strongly improve this and I think this would be easier to re-open a PR. In particular I would recommend to import CG from deepinv because at the moment your CG is not tested. 

Would you be ok to re-open a PR?
